Modules/posixmodule.c

```diff
@@ -5088,212 +5088,250 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
 }
 
 
+#define PY_IFREG 1
+#define PY_IFDIR 2
+#define PY_IFLNK 3
+#define PY_IFMNT 4
+
+static BOOL
+_testFileTypeByHandle(HANDLE hfile, int refFileType, BOOL diskOnly)
+{
+    assert(refFileType == PY_IFREG || refFileType == PY_IFDIR ||
+           refFileType == PY_IFLNK || refFileType == PY_IFMNT);
+
+    DWORD fileDevType = GetFileType(hfile);
+    if ((fileDevType == FILE_TYPE_CHAR) ||
+        (fileDevType != FILE_TYPE_DISK && diskOnly) ||
+        (fileDevType == FILE_TYPE_UNKNOWN && GetLastError() != 0))
+    {
+        return FALSE;
+    }
+
+    FILE_ATTRIBUTE_TAG_INFO atInfo;
+    FILE_BASIC_INFO baInfo;
+    DWORD attributes, reparseTag;
+    if ((refFileType == PY_IFLNK || refFileType == PY_IFMNT)) {
+        if (!GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &atInfo,
+                                          sizeof(atInfo)))
+        {
+            return FALSE;
+        }
+        attributes = atInfo.FileAttributes;
+        reparseTag = atInfo.ReparseTag;
+    }
+    else {
+        if (!GetFileInformationByHandleEx(hfile, FileBasicInfo, &baInfo,
+                                          sizeof(baInfo)))
+        {
+            return FALSE;
+        }
+        attributes = baInfo.FileAttributes;
+        reparseTag = 0;
+    }
```

Comment:
The compiler might be able to use less overall stack space if the records are declared in the inner scopes. Ideally it would just reserve enough stack space for the larger one.

```suggestion
    DWORD attributes, reparseTag;
    if ((refFileType == PY_IFLNK || refFileType == PY_IFMNT)) {
        FILE_ATTRIBUTE_TAG_INFO info;
        if (!GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &info,
                                          sizeof(info)))
        {
            return FALSE;
        }
        attributes = info.FileAttributes;
        reparseTag = info.ReparseTag;
    }
    else {
        FILE_BASIC_INFO info;
        if (!GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
                                          sizeof(info)))
        {
            return FALSE;
        }
        attributes = info.FileAttributes;
        reparseTag = 0;
    }
```